benchmark/app_test.go

```diff
@@ -0,0 +1,158 @@
+package benchmark
+
+import (
+	"strconv"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	testapp "github.com/bandprotocol/chain/v2/testing/testapp"
+	"github.com/bandprotocol/chain/v2/x/oracle/keeper"
+	oracletypes "github.com/bandprotocol/chain/v2/x/oracle/types"
+)
+
+type BenchmarkApp struct {
+	*testapp.TestingApp
+	Sender    *Account
+	Validator *Account
+	Oid       uint64
+	Did       uint64
+	TxConfig  client.TxConfig
+	TxEncoder sdk.TxEncoder
+	TB        testing.TB
+	Ctx       sdk.Context
+	Querier   keeper.Querier
+}
+
+func InitializeBenchmarkApp(b testing.TB) *BenchmarkApp {
+	ba := &BenchmarkApp{
+		TestingApp: testapp.NewTestApp("", log.NewNopLogger()),
+		Sender: &Account{
+			Account: testapp.Owner,
+			Num:     0,
+			Seq:     0,
+		},
+		Validator: &Account{
+			Account: testapp.Validators[0],
+			Num:     5,
+			Seq:     0,
+		},
+		TB: b,
+	}
+	ba.Ctx = ba.NewUncachedContext(false, tmproto.Header{})
+	ba.Querier = keeper.Querier{
+		Keeper: ba.OracleKeeper,
+	}
+	ba.TxConfig = ba.GetTxConfig()
+	ba.TxEncoder = ba.TxConfig.TxEncoder()
+
+	ba.Commit()
+	ba.CallBeginBlock()
+
+	// create oracle script
+	oCode, err := GetBenchmarkWasm()
+	require.NoError(b, err)
+	oid := ba.SendMsgCreateOracleScript(ba.Sender, oCode)
+	ba.Oid = uint64(oid)
+
+	// create data source
+	dCode := []byte("hello")
+	did := ba.SendMsgCreateDataSource(ba.Sender, dCode)
+	ba.Did = uint64(did)
+
+	// activate oracle
+	ba.SendMsgActivate(ba.Validator)
+
+	ba.CallEndBlock()
+	ba.Commit()
+
+	return ba
+}
+
+func (ba *BenchmarkApp) SendMsgCreateOracleScript(account *Account, code []byte) int {
+	tx := GenSequenceOfTxs(ba.TxConfig, GenMsgCreateOracleScript(account, code), account, 1)[0]
+	_, res, err := ba.CallDeliver(tx)
+	require.NoError(ba.TB, err)
+
+	evt := res.Events[len(res.Events)-1]
+	attr := evt.Attributes[0]
+	oid, err := strconv.Atoi(string(attr.Value))
+	require.NoError(ba.TB, err)
+
+	return oid
+}
+
+func (ba *BenchmarkApp) SendMsgCreateDataSource(account *Account, code []byte) int {
+	tx := GenSequenceOfTxs(ba.TxConfig, GenMsgCreateDataSource(account, code), account, 1)[0]
+	_, res, err := ba.CallDeliver(tx)
+	require.NoError(ba.TB, err)
+
+	evt := res.Events[len(res.Events)-1]
+	attr := evt.Attributes[0]
+	did, err := strconv.Atoi(string(attr.Value))
+	require.NoError(ba.TB, err)
+
+	return did
+}
```

Comment:
these two functions look very similar, can we refactor them?

Reply:
okay, already created new two functions (one for delivering msg, second for get value for first event)